lib/vm/src/memory.rs

```diff
@@ -135,13 +135,35 @@ pub struct LinearMemory {
     offset_guard_size: usize,
 
     /// The owned memory definition used by the generated code
-    vm_memory_definition: Box<UnsafeCell<VMMemoryDefinition>>,
+    vm_memory_definition: VMMemoryDefinitionOwnership,
 
     // Records whether we're using a bounds-checking strategy which requires
     // handlers to catch trapping accesses.
     pub(crate) needs_signal_handlers: bool,
 }
 
+/// A type to help manage who is responsible for the backing memory of them
+/// `VMMemoryDefinition`.
+#[derive(Debug)]
+enum VMMemoryDefinitionOwnership {
+    /// The `VMMemoryDefinition` is owned by the `Instance` and we should use
+    /// its memory. This is how a local memory that's exported should be stored.
+    VMOwned(NonNull<VMMemoryDefinition>),
+    /// The `VMMemoryDefinition` is owned by the host and we should manage its
+    /// memory. This is how an imported memory that doesn't come from another
+    /// Wasm module should be stored.
+    HostOwned(Box<UnsafeCell<VMMemoryDefinition>>),
+}
```

Comment:
I'm not super fan of this division. A memory is a memory, regardless of who owns it.

Reply:
Yeah, this is encapsulated internally and the reason we store it this way is so that freeing memory works correctly:

`VMMemoryDefinitionOwnership::VMOwned` won't free the memory when it gets dropped but `VMMemoryDefinitionOwnership::HostOwned` will.

I agree that it might make sense to define methods on this type to avoid having to `match` in the implementation code.

Reply:
Hypothetically if we don't fix it this way, then here's another thing we could do, not that it wouldn't be a ton of work. Make sure that VM* objects exist only inside the VMContext/Instance. wasmer_api:: would need two objects for memory, one that would be a handle to any kind of memory and the other which would be a host memory (detached from instance/vmcontext) and would contain the backing store. The VMMemoryImport would no longer point to a VMMemoryDefinition because a host Memory wouldn't have one, it would have to point to all the relevant parts (the backing, the enough to call grow and size, etc.) directly. The user has to keep the host memory and backing store alive, probably by Rc or Arc, and we could somewhat fold it into the handle (api::Memory) with an enum that indicates whether it's holding a whole backing store, or pointing to a VMMemoryDefinition.

Reply:
The idea being that `VMMemoryImport` would never own its metadata?

I'm not sure I fully understood what you're suggesting but to make sure we're on the same page:
- The generated code always needs a `VMMemoryDefinition`.
   - If it's local, it lives in VM-owned memory
   - if it's imported it lives somewhere else (either another VM's memory or the host)
- In the current code, due to `Arc`, an Instance using an imported host type will ensure that it's never freed while the Instance is using it.
- We need an extra layer of indirection for imported VM Objects to make sure that modifications are reflected in the generated code.  It's possible to remove this layer of indirection but it means we'd need update logic.

Reply:
Right, the idea being that VMMemoryImport would never own its metadata. In this model the generated code would need to change to work with a VMMemoryImport.

The compiler-generated code needs to know the base pointer and the size. It also has to be able to grow and size, both of which go through libcalls (even in the local owned memory case). More subtly we need invalidation rules for the base pointer and the size, for instance can we assume they never change at any time the function is executing?

One possible solution. You could have the VMMemoryImport have ptr_to_ versions of the fields the VMMemoryDefinition has. Instead of ptr-to-vmmemorydefinition and arc<dyn Memory> it would have a ptr_to_base and ptr_to_current_length, then the Arc<dyn Memory> (or whatever we call the object that keeps the pointed-to-memory alive, if anything) goes in the Instance::memories, not the VMContext objects.

The important part of my suggestion is to start with a rule that VM* objects are only ever part of a VMContext and then follow that through the rest of the design.

Reply:
> The important part of my suggestion is to start with a rule that VM* objects are only ever part of a VMContext and then follow that through the rest of the design.

I think this makes ownership harder in the case of a VM object imported into multiple instances, who is responsible for freeing it? What about memories created when there are no Instances created?

I'm not necessarily opposed to the changes but I think we should land this PR first, this change is self-contained and fixes user-facing bugs; the implementation can easily be changed later in my opinion.